Migration Flow Improvements: Offer 50% off for the first year on the Upgrade screen

merkushin opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #91166 

## Proposed Changes

* Remove free trial when the introductory offer is available.
* Display the introductory offer (price, badge, and CTA).
* Use hooks instead of Redux states.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* We want to offer the introductory discount to the user in the migration flow.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Enable the introductory offer feature in `config/development.json` (look for `migration-flow/introductory-offer` and set it to `true`) or use `.env` for this purpose: ```ACTIVE_FEATURE_FLAGS="migration-flow/introductory-offer"```
* Go to https://wordpress.com/me/account and switch the language to English.
* Run Calypso.
* Open `calypso.localhost:3000/start` and take the steps to reach the upgrade step (choose a domain, choose a plan, check import content in goals, choose import everything and do it on your own).
* On the upgrade step, make sure:
  * you see the discounted price and the original price next to it;
  * you see the text explaining the payment steps and taxes ("per month, $150 billed annually for the first year, $300 afterwards, excl. taxes" for USD).
  * you see the "One time offer" badge above the prices;
  * you see the CTA "Get the plan and migrate";
  * you don't see the free trial button;
  * you don't see the period switcher (annual/monthly);
  * the page looks as expected on mobile.
* Click on "Get the plan and migrate" and make sure the introductory offer is applied on the Checkout.
* Disable the feature flag (`migration-flow/introductory-offer`).
* Repeat the the steps until you get to the Upgrade step
* On the upgrade step, make sure:
  * you see the standard price;
  * CTA is "Upgrade and migrate";
  * there is the period switcher and it works as expected;
  * there is the free trial button;
  * the page looks as expected on mobile.
* Click on "Get the plan and migrate" and make sure the introductory offer is not applied on the Checkout.
* Go to https://wordpress.com/me/account and switch the language to a non-English one.
* Repeat the the steps until you get to the Upgrade step
* On the upgrade step, make sure you don't see untranslated strings.
* Change the language back to English.
* Turn off the introductory offer feature in `config/development.json`.
* Repeat the steps until you get to the upgrade step.
* On the upgrade step, make sure:
  * you see the standard price;
  * CTA is "Upgrade and migrate";
  * there is the period switcher and it works as expected;
  * there is the free trial button;
  * the page looks as expected on mobile.

## Screenshots
<img width="858" alt="Migration Flow Improvements" src="https://github.com/Automattic/wp-calypso/assets/329356/9b02ab4a-cceb-41b0-a4e8-409bb1d00c84">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

renatho committed:
Add introductory offer feature flag

merkushin committed:
Output upgrade plan details depending on the intro offer feature flag

merkushin commented:
Current state (flag enabled)
![CleanShot 2024-06-28 at 00 29 48@2x](https://github.com/Automattic/wp-calypso/assets/329356/61ca2b4c-b1c7-4b82-ba3e-21548cb23fbc)

merkushin committed:
Convert functions into components

merkushin committed:
Display introductory offer price

merkushin committed:
Get all needed data from different sources

merkushin commented:
## Current state

With intro:
![CleanShot 2024-07-02 at 01 45 44@2x](https://github.com/Automattic/wp-calypso/assets/329356/06f747c6-9341-49bd-8b3c-706b912bede8)

Without intro:
![CleanShot 2024-07-02 at 01 46 41@2x](https://github.com/Automattic/wp-calypso/assets/329356/89be54fc-146a-4fb9-a22c-da62055be30f)

merkushin commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,19 +1,219 @@
+	);
+};
+
+const preparePlanPriceOfferProps = (
+	selectedPlan: string,
+	plan?: Plan,
+	pricing?: SitePlanPricing,
+	planDetails?: SitePlanData
+): PlanPriceOfferProps => {
+	let formattedOriginalPrice = undefined;
+	if ( planDetails && planDetails.formattedOriginalPrice ) {
+		formattedOriginalPrice = planDetails.formattedOriginalPrice;
+	}
+	if ( ! formattedOriginalPrice ) {
+		formattedOriginalPrice = planDetails?.formattedPrice || undefined;
+	}
+
+	const currencyCode = pricing?.currencyCode;
+	const originalMonthlyPrice = pricing?.originalPrice.monthly || undefined;
+
+	const introOfferRawPrice = pricing?.introOffer?.rawPrice;
+	const introOfferMonthlyPrice = introOfferRawPrice
+		? calculateMonthlyPriceForPlan( selectedPlan, introOfferRawPrice )
+		: undefined;
+	const formattedIntroOfferPrice = pricing?.introOffer?.formattedPrice || undefined;
+
+	const introOfferEnabled = config.isEnabled( 'migration-flow/introductory-offer' );
+	const isIntroOfferComplete = pricing?.introOffer?.isOfferComplete || false;
+
+	return {
+		plan,
+		currencyCode,
+		originalMonthlyPrice,
+		introOfferMonthlyPrice,
+		formattedOriginalPrice,
+		formattedIntroOfferPrice,
+		introOfferEnabled,
+		isIntroOfferComplete,
+	};
+};
```
@daledupreez In case you start working on hooks.

Here I want to show that at the moment I had to use different sources (both Redux state and hooks) to gather all needed information.

I found that the hooks-approach has needed data (formatted original price) at the moment of getting response from API, but eliminates it while preparing the result.

Another interesting moment is that all sources have a mix of formats.

For example, we might have formatted price for the original price, but not for the introductory offer, or vice versa.
Or we have monthly and full price for original price and for discounted price (and they are smallest ints), but the introductory offer has only full and not the smallest int.
And so on.

I was very close to start changing the format, but realized it is a trap for me :)

daledupreez replied:
I think we've been leaning towards using the client-side `formatCurrency()` function from `@automattic/format-currency` (located in `packages/format-currency`). So I think a lot of this logic can be removed and we can simply operate on the raw prices plus the currency, and ignore the formatted values.

merkushin committed:
Merge branch 'trunk' into add/migration-flow-display-intro-offer

merkushin committed:
Show proper billing time frame

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,24 +1,220 @@
+import styled from '@emotion/styled';
 import { useI18n } from '@wordpress/react-i18n';
 import clsx from 'clsx';
+import { useTranslate } from 'i18n-calypso';
```
I think we should stick to using either `useTranslate()` or `useI18n()`, but not both.

daledupreez replied:
Noting that per our discussion, we may need the more robust string/component interpolation available via `translate()`.

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,24 +1,220 @@
 import ButtonGroup from 'calypso/components/button-group';
 import QuerySitePlans from 'calypso/components/data/query-site-plans';
 import { useSelectedPlanUpgradeMutation } from 'calypso/data/import-flow/use-selected-plan-upgrade';
+import { SitePlanData } from 'calypso/my-sites/checkout/src/hooks/product-variants';
```
I think we should import this as a type.
```suggestion
import { type SitePlanData } from 'calypso/my-sites/checkout/src/hooks/product-variants';
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,24 +1,220 @@
+	);
+};
+
+const Badge = styled.div< { isForIntroOffer?: boolean; isHidden?: boolean } >`
+	text-align: center;
+	white-space: nowrap;
+	font-size: 0.75rem;
+	line-height: 1.25rem;
+	border-radius: 4px;
+	height: 21px;
+	display: inline-block;
+	width: fit-content;
+	letter-spacing: 0.2px;
+	font-weight: 500;
+	text-align: center;
+	padding: 0 12px;
+	background-color: #e6d1ff;
+	color: #350e63;
+	text-transform: none;
+`;
```
It may be worth looking into whether it would make more sense to re-use an existing badge/pill component for this, rather than generating a completely new entity that will drift from the standard/expected behaviour over time.
How do things look if we use the `Badge` component from `@automattic/components` with the `type` attribute set to `info-purple` or `info-blue`?

daledupreez replied:
In a similar vein, it feels like we should be putting most of this CSS in the `style.scss` file in this directory (as well as possibly including that file from this component explicitly). As things stand, we have some rules in that file, and some in this one, and that's pretty messy. I can't tell you what the precedence will be, for example!

That file also has our standard variables and typography etc already pulled in, so we may be able to avoid some of the explicit colour values above.

daledupreez replied:
Also, I have some concerns about the behaviour of `height: 21px;` for folks with larger default font sizes, as well as the interplay with `white-space: no-wrap` and translated strings that may be longer than we expect. I'd rather overflow and make sense than be pixel-perfect and missing critical information.

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,24 +1,220 @@
+		introOfferEnabled && ! isIntroOfferComplete && introOfferMonthlyPrice && originalMonthlyPrice;
+
+	const price = showIntroOffer ? (
+		<PriceWithIntroductoryOffer
+			originalMonthlyPrice={ originalMonthlyPrice }
+			introOfferMonthlyPrice={ introOfferMonthlyPrice }
+			currencyCode={ currencyCode }
+		/>
```
Minor non-blocking note: I am not sure that shifting this into a new component is a good move, especially from a cognitive flow perspective, as the interior component is just returning a fragment. We could inline that and do away with the wrapper component and props interface.
```suggestion
		<>
			<Badge>{ translate( 'One time offer' ) }</Badge>
			<PricesGroup>
				<PlanPrice
					rawPrice={ originalMonthlyPrice }
					currencyCode={ currencyCode }
					original
					isSmallestUnit
				/>
				<PlanPrice rawPrice={ introOfferMonthlyPrice } currencyCode={ currencyCode } />
			</PricesGroup>
		</>
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,24 +1,220 @@
+		<PlanPrice rawPrice={ originalMonthlyPrice } currencyCode={ currencyCode } isSmallestUnit />
+	);
+
+	if ( plan && showIntroOffer && formattedIntroOfferPrice ) {
+		plan.getBillingTimeFrame = () =>
+			translate(
+				'per month, %(discounted)s billed annually for the first year, %(original)s atferwards',
+				{
+					args: {
+						discounted: formattedIntroOfferPrice,
+						original: formattedOriginalPrice || '',
+					},
+				}
+			);
+	}
```
Overwriting `plan.getBillingTimeFrame()` feels like it could easily have major unintended side-effects! I would strongly recommend we push the value into a local `const` instead.

Also, the text has a typo `atferwards` => `afterwards`, and it would be well worth adding translator comments and/or changing the variable names to describe what the values are. (The exact comment format may depend on whether you're using `__` or `translate()`, but I'll use the `translate()` format here.)
```suggestion
	const billingTimeFrame = plan && showIntroOffer && formattedIntroOfferPrice
		? translate(
			'per month, %(discountedPrice)s billed annually for the first year, %(originalPrice)s afterwards',
			{
				args: {
					discountedPrice: formattedIntroOfferPrice,
					originalPrice: formattedOriginalPrice || '',
				},
				comment: 'discountedPrice is a formatted price like $150; originalPrice is a formatted price like $200',
			}
		) : plan?.getBillingTimeFrame();		
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,24 +1,220 @@
+const UpgradePlanPeriodSwitcher = ( props: SwitcherProps ) => {
+	const { __ } = useI18n();
+
+	if ( config.isEnabled( 'migration-flow/introductory-offer' ) ) {
+		return null;
+	}
```
Hmmm... I thought we would allow users to purchase monthly plans, but would not get the introductory offer in that case, and we'd just show the standard offer. But I could easily be wrong! This just surprised me.

merkushin replied:
Yes, there were a few back and forth. 
Here is the final decision about it: p9Jlb4-c3O-p2#comment-11827

daledupreez replied:
Thanks for confirming! It will be really interesting to see if/how conversion gets impacted by removing the cheaper upfront option.

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,24 +1,220 @@
+			);
+	}
+
+	return (
+		<div className="import__upgrade-plan-price">
+			{ price }
+			<span className="plan-time-frame">
+				<small>{ plan?.getBillingTimeFrame() }</small>
+			</span>
+		</div>
+	);
+};
```
At a basic level, this component is pretty straightforward, but the code above is pretty complex. Might it be worth refactoring the core logic to build the `price` and `billingTimeFrame` functions into a local hook/function, as a way of simplifying the code? As I see it (without looking deeply), that code could return early when we're not showing the introductory offer, and simply return `<PlanPrice ... />` and `plan?.getBillingTimeFrame()`. And if we are showing the introductory offer, we can return the more complex component and the custom string for the billing time frame.

daledupreez reviewed (commented):
I've left a number of comments for you, @merkushin!

merkushin committed:
Improve pricing offer

merkushin committed:
Fix badge styling

merkushin commented:
## Current state
---

Feature enabled, introductory offer available:
![CleanShot 2024-07-02 at 22 55 02@2x](https://github.com/Automattic/wp-calypso/assets/329356/555e1e27-4cdc-481d-805c-b74ae7c7c111)
![CleanShot 2024-07-02 at 22 55 50@2x](https://github.com/Automattic/wp-calypso/assets/329356/8d3115f5-5f83-419f-bf9a-49521b4c7135)

Feature enabled, introductory offer not available:
![CleanShot 2024-07-02 at 22 56 51@2x](https://github.com/Automattic/wp-calypso/assets/329356/9cb3f6ca-7efd-494f-934a-0cc5aa9854ba)
![CleanShot 2024-07-02 at 22 56 35@2x](https://github.com/Automattic/wp-calypso/assets/329356/ac3cb2ab-5f7b-4257-ba5b-cd6615ed0fad)

Feature disabled:
![CleanShot 2024-07-02 at 23 02 39@2x](https://github.com/Automattic/wp-calypso/assets/329356/0531bed3-f60f-45a4-9cd5-807e70aa67c5)
![CleanShot 2024-07-02 at 23 02 50@2x](https://github.com/Automattic/wp-calypso/assets/329356/8c88ca6d-40cc-4a4f-ad49-4189cbfd4bc4)

merkushin commented:
Thanks for your feedback @daledupreez!
I believe I addressed all your comments and suggestions.
In this version I removed the usage of the Redux state, so we use only `useSitePlans`.
I also added the mention of taxes as we discussed in our conversation. 
Could you take another look at it?

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,31 +1,183 @@
+const UpgradePlanPeriodSwitcher = ( props: SwitcherProps ) => {
+	const translate = useTranslate();
+
+	if ( config.isEnabled( 'migration-flow/introductory-offer' ) ) {
+		return null;
+	}
```
Question: if the site is not eligible for the introductory offer, might it make sense to fall back on the original design? That way those users can still access the current UI. Phrased differently, when we enable the feature flag, we're going to hide the monthly plan from all users, even those who are not eligible for the introductory offer.

merkushin replied:
Yes, definitely makes sense. Thanks for the catch!

merkushin replied:
Addressed here: https://github.com/Automattic/wp-calypso/pull/92201/commits/5e3bfec5a15ef881fbc6f178352aab1c143754d4

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,31 +1,183 @@
+	}
+
+	const billingTimeFrame = translate(
+		'per month, %(discountedPrice)s billed annually for the first year, %(originalPrice)s afterwards, excl. taxes',
```
Minor thought: it may be clearer to specify `per year` for the final recurring price.
```suggestion
		'per month, %(discountedPrice)s billed annually for the first year, %(originalPrice)s per year afterwards, excl. taxes',
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,31 +1,183 @@
+	);
+
+	return (
+		<div className="import__upgrade-plan-price">
```
Non-blocking: this duplication of the wrapper `div` feels like it could be handled differently, maybe by pushing the logic for the price component and billing time frame into a hook. But this is a very minor quibble -- this is definitely fine as-is.

merkushin replied:
Left this one for now. Maybe address it as a follow-up.

daledupreez replied:
Still non-blocking, but this could be simplified via a lightweight component, something like:
```ts
const PlanPrice = ( { children, billingTimeFrame } ) => {
	return (
		<div className="import_upgrade-plan-price">
			{ children }
			<span className="plan-time-frame">
				<small>{ billingTimeFrame }</small>
			</span>
		</div>
	);
};
```

merkushin replied:
Extracted it here: https://github.com/Automattic/wp-calypso/pull/92201/commits/56663df41dc3513cafbeaf684cf850cd0eb5d3a1

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -1,31 +1,183 @@
+	const currencyCode = pricing?.currencyCode;
+	const originalMonthlyPrice = pricing?.originalPrice.monthly || undefined;
+
+	const introOfferRawPrice = pricing?.introOffer?.rawPrice;
+	const introOfferMonthlyPrice = introOfferRawPrice
+		? calculateMonthlyPriceForPlan( selectedPlan, introOfferRawPrice )
+		: undefined;
+
+	const originalFullPrice = pricing?.originalPrice.full || undefined;
+	const introOfferFullPrice = pricing?.introOffer?.rawPrice || undefined;
```
I don't think we need `introOfferRawPrice` here. It actually confused me a bit when reading the code.

Also, out of interest, is there a specific reason you're using `|| undefined` rather than `?? undefined`?

```suggestion
	const introOfferFullPrice = pricing?.introOffer?.rawPrice || undefined;
	const introOfferMonthlyPrice = introOfferFullPrice
		? calculateMonthlyPriceForPlan( selectedPlan, introOfferFullPrice )
		: undefined;

	const originalFullPrice = pricing?.originalPrice.full || undefined;	
```

daledupreez reviewed (commented):
I have some comments from inspection, but haven't tested yet.

merkushin committed:
Apply suggestions from code review

Co-authored-by: daledupreez <dale@automattic.com>

merkushin committed:
Use fallback translations for non-English users

- For new strings use existing fallback messages.
- Show the standard offer when the feature is enabled, but intro not
  available.
- Add tests.

merkushin committed:
Merge branch 'trunk' into add/migration-flow-display-intro-offer

merkushin commented:
## Current state
---
Intro offer available:
![CleanShot 2024-07-04 at 00 35 17@2x](https://github.com/Automattic/wp-calypso/assets/329356/1d595f24-f400-4ef9-9879-349aabbee442)
![CleanShot 2024-07-04 at 00 35 25@2x](https://github.com/Automattic/wp-calypso/assets/329356/bec827e9-2d50-4dc9-9ec6-34344465e93f)

---
Intro offer not available:
![CleanShot 2024-07-04 at 00 35 56@2x](https://github.com/Automattic/wp-calypso/assets/329356/57c89404-8339-4d9e-9cc2-46c49a888a57)
![CleanShot 2024-07-04 at 00 36 02@2x](https://github.com/Automattic/wp-calypso/assets/329356/506ffe99-dd3c-4d1b-b11b-0e7ca26c730b)

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/index.tsx:
```diff
@@ -54,7 +54,20 @@ export const UpgradePlan: React.FunctionComponent< Props > = ( props: Props ) =>
 		migrationTrialEligibility?.error_code === 'email-unverified';
 
+	const plans = Plans.useSitePlans( { siteId: site.ID } );
+	const pricing = plans.data ? plans.data[ visiblePlan ].pricing : undefined;
```
Non-blocking, but it may be worth using a conditional accessor here, just in case `plans.data[ visiblePlan ]` isn't present for some reason. It just means we'll handle the situation smoothly rather than triggering a nasty error.
```suggestion
	const pricing = plans.data ? plans.data[ visiblePlan ]?.pricing : undefined;
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/index.tsx:
```diff
@@ -54,7 +54,20 @@ export const UpgradePlan: React.FunctionComponent< Props > = ( props: Props ) =>
+		pricing.currencyCode;
+
 	const hideFreeMigrationTrial =
+		isEnabled( 'migration-flow/introductory-offer' ) ||
```
Minor thought: much like monthly plans, this will hide the trial from all users, even those not eligible for the introductory offer. Maybe we should just disable the free trial when we have an introductory offer available.
```suggestion
		introOfferAvailable ||
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/index.tsx:
```diff
@@ -82,7 +95,12 @@ export const UpgradePlan: React.FunctionComponent< Props > = ( props: Props ) =>
+			cta = translate( 'Get the plan and migrate' );
+		} else {
+			cta = cta === '' ? translate( 'Continue' ) : cta;
+		}
```
I find this code hard to follow, particularly due to `cta` being used so much.

Might it be simpler to pull some of the logic into an `else if` condition so we're splitting up the logic and the assignment?
```suggestion
		} else if ( cta === '' ) {
			cta = translate( 'Continue' );
		}
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/test/upgrade-plan-details.tsx:
```diff
@@ -0,0 +1,103 @@
+/**
+ * @jest-environment jsdom
+ */
+import { SitePlanPricing } from '@automattic/data-stores';
```
Nit: this should be a type import.
```suggestion
import { type SitePlanPricing } from '@automattic/data-stores';
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/test/upgrade-plan-details.tsx:
```diff
@@ -0,0 +1,103 @@
+	};
+}
+
+describe( 'UpgradePlanDetails', () => {
```
💖 for the tests!

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -55,31 +220,18 @@ export const UpgradePlanDetails = ( props: Props ) => {
-					</Button>
-				</ButtonGroup>
-			</div>
+			<UpgradePlanPeriodSwitcher
+				introOfferAvailable={ introOfferAvailable }
+				selectedPlan={ selectedPlan }
+				onMonthlyPlanClick={ () => setSelectedPlan( PLAN_BUSINESS_MONTHLY ) }
+				onAnnualPlanClick={ () => setSelectedPlan( PLAN_BUSINESS ) }
+			/>
```
With the change in the `UpgradePlanPeriodSwitcher` to toggle display via `introOfferAvailable`, we could simply inline that check. We could even use the old `div` from before your changes with `introOfferAvailable` as a guard.

I don't have a strong opinion, but I have a slight preference for the local guard so it's easier to understand from the main component when the toggle will (not) be shown.

```suggestion
			{ introOfferAvailable && (
				<UpgradePlanPeriodSwitcher
					introOfferAvailable={ introOfferAvailable }
					selectedPlan={ selectedPlan }
					onMonthlyPlanClick={ () => setSelectedPlan( PLAN_BUSINESS_MONTHLY ) }
					onAnnualPlanClick={ () => setSelectedPlan( PLAN_BUSINESS ) }
				/>
			)
		}
```

daledupreez commented on config/production.json:
```diff
@@ -116,8 +116,8 @@
 		"migration-flow/new-migration-instructions-step": false,
+    "migration-flow/introductory-offer": false,
 		"migration-flow/enable-migration-assistant": true,
-		"my-sites/add-ons": true,
```
I don't think we should remove this flag in this PR, especially given that it is set to `true`.

merkushin replied:
Reverted it here: https://github.com/Automattic/wp-calypso/pull/92201/commits/d36bc2571831979039ba497114ac29cc700824ff
looks like a result of an unsuccessful merging of the conflict 🤦‍♂️

daledupreez commented on config/production.json:
```diff
@@ -116,8 +116,8 @@
 		"me/account/color-scheme-picker": true,
 		"me/vat-details": true,
 		"migration-flow/new-migration-instructions-step": false,
+    "migration-flow/introductory-offer": false,
```
Nit: it looks like the formatting is off here.
```suggestion
		"migration-flow/introductory-offer": false,
```

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/style.scss:
```diff
@@ -122,6 +122,49 @@ $business-plan-color: #7f54b3;
+		}
+	}
+
+	.import__upgrade-plan-price-badge {
+		text-align: center;
+		font-size: 0.75rem;
+		line-height: 1.5;
+		height: auto;
+		border-radius: 4px;
+		display: inline-block;
+		width: fit-content;
+		letter-spacing: 0.2px;
+		font-weight: 500;
+		padding: 0 12px;
+		background-color: #e6d1ff;
+		color: #350e63;
+		text-transform: none;
+		margin-bottom: 0.5rem;
+	}
```
It feels like we're adding too much CSS here. Based on my testing, I think we'll get very similar UX if we remove the styles in my suggested edit, which also means we can pick up general changes to the Badge component and be more in sync with changes across the platform. To be clear: I am erring on the side of less long-term maintenance and more general visual consistency over pixel-perfect matching the designs. It may be worth getting input/review from the designer before making drastic changes.

```suggestion
	.import__upgrade-plan-price-badge {
		font-size: 0.75rem;
		border-radius: 4px;
		width: fit-content;
		font-weight: 500;
		background-color: #e6d1ff;
		color: #350e63;
		margin-bottom: 0.5rem;
	}
```

Also, I see that the colours don't match any colours in our default set of colour variables. I think that's non-blocking in terms of getting this shipped, _but_ I think we should double-check with the designer about possibly using colours from our palette so they are more future-proof, and revisit that pretty soon (mostly so we don't forget).

daledupreez reviewed (commented):
Overall, this is LGTM! However, I do have one blocking request to undo an accidental delete of an unrelated feature flag in `production.json`.

### Screenshots

#### Feature flag disabled

<img width="858" alt="migration-intro-offers-flag-disabled" src="https://github.com/Automattic/wp-calypso/assets/3376401/d71acf0e-b939-4476-afbc-20aaf1cbf92c">

#### Feature flag enabled; site not eligible for offer

<img width="858" alt="migration-intro-offers-flag-enabled-not-eligible" src="https://github.com/Automattic/wp-calypso/assets/3376401/b69b4025-7762-44b6-940d-8cfd066b142c">

#### Feature flag enabled; site eligible for offer

<img width="858" alt="migration-intro-offers-flag-enabled-eligible" src="https://github.com/Automattic/wp-calypso/assets/3376401/c9e079d8-de88-4cff-8b89-4a715a2dcf13">

merkushin committed:
Apply suggestions from code review

Co-authored-by: daledupreez <dale@automattic.com>

merkushin committed:
Revert removal of the flag

merkushin committed:
Extract a lightweigt price component

merkushin committed:
Update client/blocks/importer/wordpress/upgrade-plan/style.scss

Co-authored-by: daledupreez <dale@automattic.com>

daledupreez commented on client/blocks/importer/wordpress/upgrade-plan/upgrade-plan-details.tsx:
```diff
@@ -14,27 +14,22 @@ import { useHasEnTranslation } from '@automattic/i18n-utils';
 import clsx from 'clsx';
-import { useTranslate } from 'i18n-calypso';
-import React, { useState, useEffect } from 'react';
+import { TranslateResult, useTranslate } from 'i18n-calypso';
+import React, { useState, useEffect, PropsWithChildren } from 'react';
```
I think we can use type imports for `TranslateResult` and `PropsWithChildren`.
```suggestion
import { type TranslateResult, useTranslate } from 'i18n-calypso';
import React, { useState, useEffect, type PropsWithChildren } from 'react';
```

daledupreez reviewed (approved):
I've re-tested with the latest changes, and everything is looking good to me!

daledupreez commented:
As a side note for testing, you can use `&flags=someflag` as a URL parameter to enable a feature flag, and `&flags=-someflag` to disable a feature flag in any environment. So you can then test the code in the Calypso.live environment linked in [this comment](https://github.com/Automattic/wp-calypso/pull/92201#issuecomment-2196240032) without needing to (re)build the UI locally.

merkushin committed:
Import type for TranslateResult and PropsWithChildren

Co-authored-by: daledupreez <dale@automattic.com>

merkushin committed:
Adjust styling

merkushin committed:
Merge branch 'trunk' into add/migration-flow-display-intro-offer

renatho committed:
Fix skeleton styles

merkushin committed:
Update tests for upgrade plan components

merkushin committed:
Fix test for upgrade plan without hoc

chriskmnds commented on client/blocks/importer/wordpress/upgrade-plan/index.tsx:
```diff
@@ -54,7 +41,23 @@ export const UpgradePlan: React.FunctionComponent< Props > = ( props: Props ) =>
+	const { list: upgradePlanHostingDetailsList, isFetching: isFetchingHostingDetails } =
+		useUpgradePlanHostingDetailsList();
+
+	const plans = Plans.useSitePlans( { siteId: site.ID } );
```
Using `usePricingMetaForGridPlans` for pricing instead of the query hook is probably better for consistency. It will also fix the issue with always showing prorated prices. There is a PR that makes it optional.

Unless there is a specific other exception, then `usePricingMetaForGridPlans` with `siteId` should give the equivalent of `useSitePlans`.

merkushin replied:
Thanks for the suggestion, updated here: https://github.com/Automattic/wp-calypso/pull/92201/commits/e4b21382979fbc2d9898892ce5eda5ccaee6f58e

I chose `useSitePlans`, because the call looked less wordy :) `usePricingMetaForGridPlans` has other required properties that makes the call spread on several lines.

chriskmnds reviewed (commented):
This is not a direct feedback to work done here, but rather more general feedback on the direction taken.

I find it a little surprising that we have such similar card design to the main `/plans` grid, yet nothing shared between. The billing descriptions are redone, the header prices are redone, the CTA buttons are redone, the features list is redone, etc. etc. Everything feels like redone for the most part. 😞 

We have been trying to make all of these components/sections generic in `@automattic/plans-grid-next`, so I would like to hope it wouldn't have been too much work to repurpose some of that instead of recreating, irrespective oh how more tailored or complex it is for the perspective use case. Perhaps exporting a mobile plan card, or some of the invidivual/granular components or hooks - or at least make things into a suggestion for "hey we want to reuse but can't for XYZ reasons". At the very least things like billing descriptions, header prices, features list, could/should be exported and reused from `@automattic/plans-grid-next`

Please don't take my feedback as judgment. Re-implementing "anything" is a cause for inconsistent look & feel across the codebase, more so unnecessary efforts going into maintenance. We have similar issue with the current/spotlight plan cards ( pdvytD-Fm-p2 ). This is now another case of the mobile plan cards.

Sorry if not looking deeper or investigating more. On the surface, I've seen surprisingly similar design and new hooks/componets serving that design from `client/blocks/importer/wordpress/upgrade-plan/*` 🤔 Hopefully a multi-plan display next won't lead to a new grid implementation under same folder

I think you all are doing great work - so, again, please don't take my feedback as judgment.

@daledupreez @Automattic/martech-decepticons

renatho commented on client/blocks/importer/wordpress/upgrade-plan/index.tsx:
```diff
@@ -2,34 +2,21 @@ import { recordTracksEvent } from '@automattic/calypso-analytics';
-}
-
-export const UpgradePlan: React.FunctionComponent< Props > = ( props: Props ) => {
+export const UpgradePlanUnwrapped: React.FunctionComponent< UpgradePlanProps > = ( props ) => {
```
Nice name solution you used here!

Nit: I'd suggest to use the "Unwrapped" as a prefix, just to use the same standard we used in the HOC (`WrappedComponent`).

merkushin replied:
Updated here: https://github.com/Automattic/wp-calypso/pull/92201/commits/e4b21382979fbc2d9898892ce5eda5ccaee6f58e

renatho commented on client/blocks/importer/wordpress/upgrade-plan/index.tsx:
```diff
@@ -137,6 +145,10 @@ export const UpgradePlan: React.FunctionComponent< Props > = ( props: Props ) =>
 				'Migrations are exclusive to the Creator plan. Check out all its benefits, and upgrade to get started.'
 		  );
 
+	if ( isFetchingHostingDetails ) {
```
Should we also test if the plans are loading?

In my tests, I didn't see anything weird, but maybe it's because that usually the plans are coming faster than the hosting details?

renatho replied:
As we talked in Slack, based on [this](https://github.com/Automattic/wp-calypso/blob/54e8be25a5056f8f1b709ff93139de03037ad861/packages/data-stores/src/plans/hooks/use-pricing-meta-for-grid-plans.ts#L107-L113), it seems fine to check it as `! pricing`, as you did in https://github.com/Automattic/wp-calypso/pull/92201/commits/850481c55c5be4335e48bf0b20788fe6cc36a62c.

But maybe @chriskmnds could double-check to confirm it? :)

merkushin replied:
Checking it here: https://github.com/Automattic/wp-calypso/pull/92201/commits/850481c55c5be4335e48bf0b20788fe6cc36a62c
The test added here: https://github.com/Automattic/wp-calypso/pull/92201/commits/9a1f0013f4c97c4b63b16ecc7f5d90290a5a51e3

merkushin replied:
> But maybe @chriskmnds could double-check to confirm it? :)

I think it is not critical and shouldn't block this PR, we can address feedback in a follow-up PR if any.

renatho commented on client/blocks/importer/wordpress/upgrade-plan/test/index.tsx:
```diff
@@ -56,11 +45,52 @@ const DEFAULT_SITE_CAPABILITIES = {
 };
 
+// Stub out UpgradePlanDetails because it has much more complex dependencies, and only provides a wrapper around the content from this component.
+jest.mock( '../upgrade-plan-details', () => ( {
```
Nit:
Was there a special reason to move the mocks after the constants and add the other mocks before the constants? Could we keep one block of code for mocks and another for the constants to make it more organized?

merkushin replied:
Updated here: https://github.com/Automattic/wp-calypso/pull/92201/commits/edff35cbaf5b91c6ee6cfa3275d5a17b3d2ad7c0

renatho commented on client/blocks/importer/wordpress/upgrade-plan/test/index.tsx:
```diff
@@ -2,37 +2,26 @@
 import { renderWithProvider } from 'calypso/test-helpers/testing-library';
-import { UpgradePlan } from '../index';
+import { useUpgradePlanHostingDetailsList } from '../hooks/use-get-upgrade-plan-hosting-details-list';
+import { UpgradePlan, UpgradePlanUnwrapped as UpgradePlanWithoutHOC } from '../index';
```
Nit: I'd say that with the new name you used, we wouldn't need the `as` here.

merkushin replied:
Updated here: https://github.com/Automattic/wp-calypso/pull/92201/commits/e4b21382979fbc2d9898892ce5eda5ccaee6f58e

renatho reviewed (commented):
@merkushin, I took a look at the changes related to the rebase. I added some small suggestions and a question about the loading status to see if we need to tweak it.

Thank you for working on this!

merkushin committed:
Replace useSitePlans with usePricingMetaForGridPlans, rename unwrapped component

merkushin committed:
Mock data-stores to fix tests

merkushin committed:
Check pricing was loaded

merkushin committed:
Add a test for the case pricing is not available

merkushin committed:
Adding mocks in tests

merkushin committed:
Fix tests for site migration upgrade plan

merkushin committed:
Fix tests in pre-migration

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16133484

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Hi @merkushin, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Get the plan and migrate`
- `per month, %(discountedPrice)s billed annually for the first year, %(originalPrice)s per year afterwards, excl. taxes`
- `One time offer`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16133484).
